### PR TITLE
Extending the Easy Code Touch

### PR DESCRIPTION
**Pull request by OsteHovel**
Adding pin code programming, presenting action and auto relock.

### CODE REVIEW

**Comment by OsteHovel** [comment]
[Easy Access E-Life Zigbee Module.pdf](https://github.com/Koenkk/zigbee-herdsman-converters/files/7229993/Easy.Access.E-Life.Zigbee.Module.pdf)
Manual ;-)

File: converters/fromZigbee.js [hunk 1]
```diff
@@ -2488,6 +2488,17 @@ const converters = {
             }
         },
     },
+    easycodetouch_action: {
+        cluster: 'closuresDoorLock',
+        type: 'raw',
+        convert: (model, msg, publish, options, meta) => {
+            const value = constants.easyCodeTouchActions[(msg.data[3] << 8) | msg.data[4]];
+            if (!value) {
+                throw new Error('Unknown lock status with source ' + msg.data[3] + ' and event code ' + msg.data[4]);
```
Comment (Koenkk) [inline converters/fromZigbee.js:2497]: Instead of throwing an error, please do ` debug.log`

File: converters/fromZigbee.js [hunk 2]
```diff
@@ -2488,6 +2488,17 @@ const converters = {
             }
         },
     },
+    easycodetouch_action: {
+        cluster: 'closuresDoorLock',
+        type: 'raw',
```
Comment (Koenkk) [inline converters/fromZigbee.js:2493]: I'm wondering why `raw` is needed here (it means  parsing the package failed). Could you provide the herdsman debug log of just this action?

See https://www.zigbee2mqtt.io/information/debug.html#zigbee-herdsman-debug-logging on how to enable the herdsman debug logging. Note that this is only logged to STDOUT and not to log files.

Comment (OsteHovel) [reply converters/fromZigbee.js:2493]:
```
zigbee_1        | Zigbee2MQTT:debug 2021-09-25 19:30:43: Received Zigbee message from 'Entrance door', type 'attributeReport', cluster 'closuresDoorLock', data '{"256":131072}' from endpoint 11 with groupID null
zigbee_1        | 2021-09-25T17:30:43.411Z zigbee-herdsman:controller:endpoint DefaultResponse 0xf4ce36516edd072e/11 257(10, {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":1,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false})
zigbee_1        | 2021-09-25T17:30:43.411Z zigbee-herdsman:deconz:adapter zclFrame.payload:
zigbee_1        | 2021-09-25T17:30:43.412Z zigbee-herdsman:deconz:adapter { cmdId: 10, statusCode: 0 }
zigbee_1        | 2021-09-25T17:30:43.412Z zigbee-herdsman:deconz:adapter no response expected
zigbee_1        | 2021-09-25T17:30:43.412Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
zigbee_1        | 2021-09-25T17:30:43.414Z zigbee-herdsman:deconz:driver DATA_REQUEST - destAddr: 0x6fb5 EP:11 SeqNr. 60 request id: 14
zigbee_1        | 2021-09-25T17:30:43.424Z zigbee-herdsman:deconz:frameParser DATA_REQUEST RESPONSE - request id: 14 status: 0
zigbee_1        | 2021-09-25T17:30:43.424Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
zigbee_1        | 2021-09-25T17:30:43.424Z zigbee-herdsman:deconz:frameParser DEVICE_STATE changed: 10100110
zigbee_1        | 2021-09-25T17:30:43.424Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 1 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
zigbee_1        | 2021-09-25T17:30:43.429Z zigbee-herdsman:deconz:driver query aps data confirm
zigbee_1        | 2021-09-25T17:30:43.435Z zigbee-herdsman:deconz:driver DATA_CONFIRM - sending data state request - SeqNr. 61
zigbee_1        | 2021-09-25T17:30:43.437Z zigbee-herdsman:deconz:frameParser DATA_CONFIRM RESPONSE - destAddr: 0x6fb5 request id: 14 confirm status: 0
zigbee_1        | 2021-09-25T17:30:43.437Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
zigbee_1        | 2021-09-25T17:30:43.437Z zigbee-herdsman:deconz:adapter sendZclFrameToEndpoint - message send with transSeq Nr.: 131
zigbee_1        | 2021-09-25T17:30:43.638Z zigbee-herdsman:deconz:frameParser DEVICE_STATE changed: 10101010
zigbee_1        | 2021-09-25T17:30:43.638Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 1 configChanged: 0 apsRequestFreeSlots: 1
zigbee_1        | 2021-09-25T17:30:43.642Z zigbee-herdsman:deconz:driver query aps data indication
zigbee_1        | 2021-09-25T17:30:43.648Z zigbee-herdsman:deconz:driver DATA_INDICATION - sending read data request - SeqNr. 62
zigbee_1        | 2021-09-25T17:30:43.650Z zigbee-herdsman:deconz:frameParser DATA_INDICATION RESPONSE - seqNr. 62 srcAddr: 0x6fb5 destAddr: 0x0 profile id: 0x104 cluster id: 0x101 lqi: 255
zigbee_1        | 2021-09-25T17:30:43.650Z zigbee-herdsman:deconz:frameParser response payload: 25,132,32,255,14,0,0,0,0,0,0
zigbee_1        | 2021-09-25T17:30:43.653Z zigbee-herdsman:deconz:adapter could not parse zclFrame: RangeError [ERR_OUT_OF_RANGE]: The value of "offset" is out of range. It must be >= 0 and <= 7. Received 8
zigbee_1        | 2021-09-25T17:30:43.655Z zigbee-herdsman:controller:log Received 'raw' data '{"clusterID":257,"data":{"type":"Buffer","data":[25,132,32,255,14,0,0,0,0,0,0]},"address":28597,"endpoint":11,"linkquality":255,"groupID":null,"wasBroadcast":false,"destinationEndpoint":1}'
zigbee_1        | Zigbee2MQTT:debug 2021-09-25 19:30:43: Received Zigbee message from 'Entrance door', type 'raw', cluster 'closuresDoorLock', data '{"data":[25,132,32,255,14,0,0,0,0,0,0],"type":"Buffer"}' from endpoint 11 with groupID null
zigbee_1        | Zigbee2MQTT:debug 2021-09-25 19:30:43: No converter available for 'easyCodeTouch_v1' with cluster 'closuresDoorLock' and type 'raw' and data '{"data":[25,132,32,255,14,0,0,0,0,0,0],"type":"Buffer"}'
zigbee_1        | 2021-09-25T17:30:43.658Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
zigbee_1        | 2021-09-25T17:30:44.030Z zigbee-herdsman:deconz:driver send read device state from queue. seqNr: 63
zigbee_1        | 2021-09-25T17:30:44.033Z zigbee-herdsman:deconz:frameParser device state: 10100010
zigbee_1        | 2021-09-25T17:30:44.033Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
zigbee_1        | 2021-09-25T17:30:44.413Z zigbee-herdsman:deconz:frameParser DEVICE_STATE changed: 10101010
zigbee_1        | 2021-09-25T17:30:44.413Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 1 configChanged: 0 apsRequestFreeSlots: 1
zigbee_1        | 2021-09-25T17:30:44.414Z zigbee-herdsman:deconz:driver query aps data indication
zigbee_1        | 2021-09-25T17:30:44.420Z zigbee-herdsman:deconz:driver DATA_INDICATION - sending read data request - SeqNr. 64
zigbee_1        | 2021-09-25T17:30:44.422Z zigbee-herdsman:deconz:frameParser DATA_INDICATION RESPONSE - seqNr. 64 srcAddr: 0x6fb5 destAddr: 0x0 profile id: 0x104 cluster id: 0x101 lqi: 255
zigbee_1        | 2021-09-25T17:30:44.422Z zigbee-herdsman:deconz:frameParser response payload: 25,134,32,255,14,0,0,0,0,0,0
zigbee_1        | 2021-09-25T17:30:44.425Z zigbee-herdsman:deconz:adapter could not parse zclFrame: RangeError [ERR_OUT_OF_RANGE]: The value of "offset" is out of range. It must be >= 0 and <= 7. Received 8
zigbee_1        | 2021-09-25T17:30:44.427Z zigbee-herdsman:controller:log Received 'raw' data '{"clusterID":257,"data":{"type":"Buffer","data":[25,134,32,255,14,0,0,0,0,0,0]},"address":28597,"endpoint":11,"linkquality":255,"groupID":null,"wasBroadcast":false,"destinationEndpoint":1}'
zigbee_1        | Zigbee2MQTT:debug 2021-09-25 19:30:44: Received Zigbee message from 'Entrance door', type 'raw', cluster 'closuresDoorLock', data '{"data":[25,134,32,255,14,0,0,0,0,0,0],"type":"Buffer"}' from endpoint 11 with groupID null
zigbee_1        | Zigbee2MQTT:debug 2021-09-25 19:30:44: No converter available for 'easyCodeTouch_v1' with cluster 'closuresDoorLock' and type 'raw' and data '{"data":[25,134,32,255,14,0,0,0,0,0,0],"type":"Buffer"}'
zigbee_1        | 2021-09-25T17:30:44.431Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1

```

File: converters/fromZigbee.js (hunk 1 above)
Comment (OsteHovel) [reply converters/fromZigbee.js]: Now using meta.logger.log

File: converters/fromZigbee.js (hunk 2 above)
Comment (Koenkk) [reply converters/fromZigbee.js:2493]: I believe it is indeed a corrupt command, do you by any change have a CC2531 to sniff the traffic when this command gets send? https://www.zigbee2mqtt.io/how_tos/how_to_sniff_zigbee_traffic.html 

With wireshark we can confirm the device sends malformed messages.

Comment (OsteHovel) [reply converters/fromZigbee.js:2493]: I will sniff some traffic this weekend 👍

Comment (OsteHovel) [reply converters/fromZigbee.js:2493]: image
It does indeed send weird data

Key: 0C:9F:45:81:8F:14:A9:53:D0:99:3F:23:04:01:5E:10
Dump: https://cloud.ostsoft.com/s/LpeS3aLoZEo8pjF

Comment (Koenkk) [reply converters/fromZigbee.js:2493]: Thanks for checking! Then we indeed need this custom converter.

**Pull request merged**